kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java

```diff
@@ -0,0 +1,503 @@
+   * @return a statement that, when executed, will insert the value
+   * for this key and partition into the table.
+   * Note that the key in this case is a "session" key, where
+   * {@code sessionKey} includes both the record key and
+   * the session start / end timestamps
+   */
+  @Override
+  public WriteModel<SessionDoc> insert(
+      final int kafkaPartition,
+      final SessionKey sessionKey,
+      final byte[] value,
+      final long epochMillis
+  ) {
+    final var partitionSegments = kafkaPartitionToSegments.get(kafkaPartition);
+
+    final long epoch = partitionSegments.epoch;
+    return new UpdateOneModel<>(
+        Filters.and(
+            Filters.eq(SessionDoc.ID, compositeKey(sessionKey)),
+            Filters.lte(SessionDoc.EPOCH, epoch)
+        ),
+        Updates.combine(
+            Updates.set(SessionDoc.VALUE, value),
+            Updates.set(SessionDoc.EPOCH, epoch),
+            Updates.unset(SessionDoc.TOMBSTONE_TS)
+        ),
+        UPSERT_OPTIONS
+    );
+  }
+
+  /**
+   * @param kafkaPartition the kafka partition
+   * @param sessionKey     the session data key
+   * @return a statement that, when executed, will delete the value
+   * for this key and partition in the table.
+   * Note that the key in this case is a "session" key, where
+   * {@code sessionKey} includes both the record key and
+   * the sessionStart timestamp
+   */
+  @Override
+  public WriteModel<SessionDoc> delete(
+      final int kafkaPartition,
+      final SessionKey sessionKey
+  ) {
+    final long epoch = kafkaPartitionToEpoch.get(kafkaPartition);
+    return new UpdateOneModel<>(
+        Filters.and(
+            Filters.eq(SessionDoc.ID, compositeKey(sessionKey)),
+            Filters.lte(SessionDoc.EPOCH, epoch)
+        ),
+        Updates.combine(
+            Updates.unset(SessionDoc.VALUE),
+            Updates.set(SessionDoc.TOMBSTONE_TS, Date.from(Instant.now())),
+            Updates.set(SessionDoc.EPOCH, epoch)
+        ),
+        UPSERT_OPTIONS
+    );
+  }
+
+  @Override
+  public byte[] fetch(
+      final int kafkaPartition,
+      final Bytes key,
+      final long sessionStart,
+      final long sessionEnd
+  ) {
+    final SessionKey sessionKey = new SessionKey(key, sessionStart, sessionEnd);
+    final Segmenter.SegmentPartition segment =
+        partitioner.tablePartition(kafkaPartition, sessionKey);
+    final var segmentSessions = sessionsForSegmentPartition(kafkaPartition, segment);
+    if (segmentSessions == null) {
+      return null;
+    }
+
+    final SessionDoc sessionDoc = segmentSessions.find(
+            Filters.and(
+                Filters.eq(SessionDoc.ID, compositeKey(sessionKey))
+            ))
+        .first();
+
+    return sessionDoc == null ? null : sessionDoc.value();
+  }
+
+  public KeyValueIterator<SessionKey, byte[]> fetchAll(
+      final int kafkaPartition,
+      final Bytes key,
+      final long earliestSessionEnd,
+      final long latestSessionEnd
```

Comment:
```suggestion
      final long latestSessionStart
```

ditto elsewhere this appears

Reply:
I am actually using earliest / latest session end to filter the remote table and find the sessions.